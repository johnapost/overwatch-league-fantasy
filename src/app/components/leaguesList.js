// @flow

import React from "react";
import { compose } from "redux";
import { connect } from "react-redux";
import { Modal, Button, Divider, Form, Input } from "antd";
import CreateLeague from "./createLeague";
import withFirestore from "../shared/withFirestore";

import type { StoreState } from "../shared/makeStore";

type Props = {
  leagues: Object[],
  user: {
    displayName: string
  }
};

export const LeaguesListComponent = ({
  user: { displayName },
  leagues
}: Props) => (
  <div>
    {leagues.length ? (
      <div>{JSON.stringify(leagues)}</div>
    ) : (
      <Modal
        title={`Welcome, ${displayName}`}
        visible
        footer={null}
        keyboard={false}
        autoFocusButton={null}
        closable={false}
      >
        <p>Join an existing league or create your own!</p>
        <Form layout="inline">
          <Form.Item>
            <Input placeholder="Invite code" disabled />
          </Form.Item>
          <Form.Item>
            <Button type="primary" disabled>
              Join someone&rsquo;s league
            </Button>
          </Form.Item>
        </Form>
        <Divider />
        <CreateLeague />
      </Modal>
    )}
    <style jsx>{`
      p {
        margin: 10px 0 30px;
      }
    `}</style>
  </div>
);

const filterMyLeagues = (
  uid: string | null,
  leagues: Object
): [string, Object][] =>
  (uid &&
    leagues &&
    ((Object.entries(leagues): any): Object[]).reduce(
      (accum, [key, value]) =>
        value.leagueUsers.includes(uid) ? [...accum, [key, value]] : accum,
      []
    )) ||
  [];

export const mapStateToProps = ({
  firestore: {
    data: { leagues }
  },
  user: { displayName, uid }
}: StoreState) => ({
  leagues: filterMyLeagues(uid, leagues),
  user: {
    displayName,
    uid
  }
});

export default compose(
  connect(mapStateToProps),
  withFirestore(({ user: { uid } }) =>
    uid
      ? [
          {
            collection: "leagues",
            where: [["leagueUsers", "array-contains", uid]]
          }
        ]
      : []
  )
)(LeaguesListComponent);
