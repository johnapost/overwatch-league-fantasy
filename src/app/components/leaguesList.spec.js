import React from "react";
import { shallow } from "enzyme";
import { Modal } from "antd";
import { LeaguesListComponent, mapStateToProps } from "./leaguesList";

const props = {
  leagues: {},
  user: {
    displayName: "John again"
  }
};

describe("FindLeague", () => {
  describe("component", () => {
    const wrapper = shallow(<LeaguesListComponent {...props} />);

    describe("with no leagues", () => {
      it("should render Modal with user displayName", () => {
        expect(wrapper.find(Modal).length).toBe(1);
        expect(wrapper.find(Modal).prop("title")).toMatch(
          props.user.displayName
        );
      });
    });

    describe("with leagues", () => {
      xit("should render a list of leagues", () => {});
    });
  });

  describe("mapStateToProps", () => {
    it("should assign user data", () => {
      expect(
        mapStateToProps({
          firestore: { data: {} },
          user: { displayName: "user", uid: "abc123" }
        })
      ).toMatchSnapshot();
    });

    it("should filter and map leagues", () => {
      expect(
        mapStateToProps({
          firestore: {
            data: {
              leagues: {
                1: { leagueUsers: "abc123" },
                2: { leagueUsers: "cba321" },
                3: { leagueUsers: "abc123" }
              }
            }
          },
          user: { uid: "abc123" }
        })
      ).toMatchSnapshot();
    });
  });
});
