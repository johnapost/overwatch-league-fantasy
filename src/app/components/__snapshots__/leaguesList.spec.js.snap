// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindLeague mapStateToProps should assign user data 1`] = `
Object {
  "leagues": Array [],
  "user": Object {
    "displayName": "user",
    "uid": "abc123",
  },
}
`;

exports[`FindLeague mapStateToProps should filter and map leagues 1`] = `
Object {
  "leagues": Array [
    Array [
      "1",
      Object {
        "leagueUsers": "abc123",
      },
    ],
    Array [
      "3",
      Object {
        "leagueUsers": "abc123",
      },
    ],
  ],
  "user": Object {
    "displayName": undefined,
    "uid": "abc123",
  },
}
`;
